docs/DafnyRef/Topics.md

```diff
@@ -1025,44 +1025,90 @@ Folks, it doesn't get any simpler than that!
-initialized. This rule ensures that any method of the class (which does not
-know which constructor may have been already called) can rely on the fields
-having been initialized.
-
-[This document](../Compilation/AutoInitialization) has more detail on
-auto-initialization.
+The Dafny language semantics ensures that any use (read) of a variable (or constant,
+parameter, object field, or array element) gives a value of the variable's type.
+It is easy to see that this property holds for any variable that is declared with
+an initializing assignment. However, for many useful programs, it would be too strict
+to require an initializing assignment at the time a variable is declared.
+Instead, Dafny ensures the property through _auto-initialization_ and rules for _definite assignment_.
+
+As explained in [section 5.3.1](#sec-type-characteristics), each type in Dafny is one of the following:
+
+- _auto-init type_: the type is nonempty and the compiler has some way to emit code that constructs a value
+- _nonempty type_: the type is nonempty, but the compiler does not know how perform automatic initialization
+- _possibly empty type_: the type is not known for sure to have a value
+
+For a variable of an auto-init type, the compiler can initialize the variable automatically.
+This means that the variable can be used immediately after declaration, even if the program does not
+explicitly provide an initializing assignment.
+
+In a ghost context, one can an imagine a "ghost" that initializes variables. Unlike the compiler, such
+a "ghost" does not need to emit code that constructs an initializing value; it suffices for the ghost to
+know that a value exists. Therefore, in a ghost context, a variable of a nonempty type can be used immediately
+after declaration.
+
+Before a variable of a possibly empty type can be used, the program must initialize it.
+The variable need not be given a value when it is declared,
+but it must have a value by the time it is first used. Dafny uses the precision of the verifier to
+reason about the control flow between assignments and uses of variables, and it reports an error
+if it cannot assure itself that the variable has been given a value.
+
+The elements of an array must be assured to have values already in the statement that allocates the array.
+This is achieved in any of the following four ways:
+- If the array is allocated to be empty (that is, one of its dimensions is requested to be 0), then
+  the array allocation trivially satisfies the requirement.
+- If the element type of the array is an auto-init type, then nothing further is required by the program.
+- If the array allocation occurs in a ghost context and the element type is a nonempty type, then nothing
+  further is required by the program.
+- Otherwise, the array allocation must provide an initialization display or an initialization function.
+See [section 0.0.0](#sec-array-type) for information about array initialization.
```

Comment:
```suggestion
See [section 5.10](#sec-array-type) for information about array initialization.
```

Reply:
Oops- rerunning the auto-numbering script slipped my mind...